File: 20-ValidParentheses/20-ValidParentheses.java
// Last updated: 8/11/2026, 12:27:09 PM
 class Solution {
    public boolean isValid(String s) {
   char[] stack = new char[s.length()];
   int top = -1;

  for (int i = 0; i < s.length(); i++) {
char c = s.charAt(i);


 if (c == '(' || c == '{' || c == '[') {
 if (top == s.length() - 1) {
              
 return false;
 }
   stack[++top] = c;
 } else {
                
   if (top == -1) {
  return false; 
 }
char last = stack[top--];

 if (!isPair(last, c)) {
  return false;
  }
   }
    }
 
  return top == -1;
    }

    private boolean isPair(char last, char c) {
        return (last == '(' && c == ')') ||
               (last == '{' && c == '}') ||
               (last == '[' && c == ']');
    }
} 

File: 20-ValidParentheses/20-ValidParentheses.py
# Last updated: 9/2/2026, 12:45:40 PM
class Solution:
    def isValid(self, s):
        stack = []
        pairs = {
            ')': '(',
            ']': '[',
            '}': '{'
        }

        for ch in s:
            if ch in pairs:
                if not stack or stack.pop() != pairs[ch]:
                    return False
            else:
                stack.append(ch)

        return not stack